be/src/storage/task/engine_storage_migration_task.cpp

```diff
@@ -161,57 +144,124 @@ OLAPStatus EngineStorageMigrationTask::_storage_medium_migrate(TTabletId tablet_
+            tablet->set_is_migrating(false);
             break;
         }
-        tablet->obtain_header_rdlock();
-        _generate_new_header(stores[0], shard, tablet, consistent_rowsets, new_tablet_meta);
-        tablet->release_header_lock();
-        std::string new_meta_file = schema_hash_path + "/" + std::to_string(tablet_id) + ".hdr";
+
+        // 3. lock and check again, and load new tablet from dest dir
+        std::unique_lock migration_wlock(tablet->get_migration_lock());
+        DeferOp reset_tablet_is_migrating([&tablet] { return tablet->set_is_migrating(false); });
+
+        std::lock_guard push_lock(tablet->get_push_lock());
+
+        int64_t partition_id;
+        std::set<int64_t> transaction_ids;
+        StorageEngine::instance()->txn_manager()->get_tablet_related_txns(
+                _tablet_id, _schema_hash, tablet->tablet_uid(), &partition_id, &transaction_ids);
+        if (!transaction_ids.empty()) {
+            LOG(WARNING) << "could not migration because has unfinished txns.";
+            need_remove_new_path = true;
+            res = OLAP_ERR_HEADER_HAS_PENDING_DATA;
+            break;
+        }
+
+        auto mem_tracker = std::make_unique<MemTracker>();
+        TabletMetaSharedPtr new_tablet_meta(new (std::nothrow) TabletMeta(mem_tracker.get()));
+        Status st = TabletMetaManager::get_tablet_meta(_dest_store, _tablet_id, _schema_hash, new_tablet_meta);
+        if (st.ok()) {
+            LOG(WARNING) << "tablet_meta already exists. data_dir:" << _dest_store->path()
+                         << "tablet:" << tablet->full_name();
+            need_remove_new_path = true;
+            res = OLAP_ERR_META_ALREADY_EXIST;
+            break;
+        } else if (!st.is_not_found()) {
+            LOG(WARNING) << "fail to get tablet meta: " << st;
+            need_remove_new_path = true;
+            res = OLAP_ERR_IO_ERROR;
+            break;
+        }
+
+        {
+            // check version
+            std::shared_lock header_rdlock(tablet->get_header_lock());
+            const RowsetSharedPtr lastest_version = tablet->rowset_with_max_version();
+            if (lastest_version == nullptr) {
+                LOG(WARNING) << "tablet has no any version.";
+                need_remove_new_path = true;
+                res = OLAP_ERR_VERSION_NOT_EXIST;
+                break;
+            }
+            int32_t new_end_version = lastest_version->end_version();
+            if (end_version != new_end_version) {
+                LOG(WARNING) << "copied version=" << end_version
+                             << " is different from new version=" << new_end_version;
+                need_remove_new_path = true;
+                res = OLAP_ERR_VERSION_NOT_EXIST;
+                break;
+            }
+
+            // generate new header
+            _generate_new_header(_dest_store, shard, tablet, consistent_rowsets, new_tablet_meta);
```

Comment:
Does primary key table support this feature?

Reply:
Primary key tablet migration is disabled and will be supported in the feature